Divide function clause building in a single module per encoding

hawkyre opened the pull request:
Fixes https://github.com/tallakt/codepagex/issues/33

This will improve compilation memory usage heavily, since instead of loading all function clauses into a single module, we only need to have each encoding's function clauses in memory at once. So now, the max memory usage will be the largest encoding requested by the user instead of the sum of all encodings.

hawkyre commented:
Running mix compile followed by mix release does create the release with the dynamic module files packed up, see the attached image. Feel free to test it locally @gitneko 
![image](https://github.com/user-attachments/assets/f0298f12-940e-41a9-90ac-f948de7ec6b8)

Fritzelblitz commented:
I'm afraid your solution pessimizes the performance and especially the memory consumption for very large strings.
(Interestingly, the benchmark for small strings is better than in the main).

Main:
```
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
Number of Available Cores: 12
Available memory: 22.73 GB
Elixir 1.18.3
Erlang 27.3.3
JIT enabled: true

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 3 s
memory time: 2 s
reduction time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 49 s

Benchmarking ascii_from_gigantic_string ...
Benchmarking ascii_from_string ...
Benchmarking ascii_to_string ...
Benchmarking erlang_unicode_from_gigantic_string ...
Benchmarking iso_from_gigantic_string ...
Benchmarking iso_from_string ...
Benchmarking iso_to_string ...
Calculating statistics...
Formatting results...

Name                                          ips        average  deviation         median         99th %
ascii_to_string                           5810.80      0.00017 s    ±29.26%      0.00016 s      0.00035 s
iso_to_string                             4767.69      0.00021 s    ±33.55%      0.00019 s      0.00044 s
iso_from_string                            788.89      0.00127 s    ±20.58%      0.00123 s      0.00196 s
ascii_from_string                          261.97      0.00382 s    ±13.91%      0.00371 s      0.00543 s
ascii_from_gigantic_string                 0.0922        10.85 s     ±0.00%        10.85 s        10.85 s
iso_from_gigantic_string                   0.0735        13.61 s     ±0.00%        13.61 s        13.61 s
erlang_unicode_from_gigantic_string        0.0468        21.39 s     ±0.00%        21.39 s        21.39 s

Comparison: 
ascii_to_string                           5810.80
iso_to_string                             4767.69 - 1.22x slower +0.00004 s
iso_from_string                            788.89 - 7.37x slower +0.00110 s
ascii_from_string                          261.97 - 22.18x slower +0.00365 s
ascii_from_gigantic_string                 0.0922 - 63033.45x slower +10.85 s
iso_from_gigantic_string                   0.0735 - 79079.60x slower +13.61 s
erlang_unicode_from_gigantic_string        0.0468 - 124287.00x slower +21.39 s

Memory usage statistics:

Name                                   Memory usage
ascii_to_string                              125 KB
iso_to_string                                125 KB - 1.00x memory usage +0 KB
iso_from_string                           195.31 KB - 1.56x memory usage +70.31 KB
ascii_from_string                        6757.81 KB - 54.06x memory usage +6632.81 KB
ascii_from_gigantic_string                414.06 KB - 3.31x memory usage +289.06 KB
iso_from_gigantic_string                  195.31 KB - 1.56x memory usage +70.31 KB
erlang_unicode_from_gigantic_string  15661817.61 KB - 125294.54x memory usage +15661692.61 KB

**All measurements for memory usage were the same**
```
Your Branch:
```
Operating System: Linux
CPU Information: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz
Number of Available Cores: 12
Available memory: 22.73 GB
Elixir 1.18.3
Erlang 27.3.3
JIT enabled: true

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 3 s
memory time: 2 s
reduction time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 49 s

Benchmarking ascii_from_gigantic_string ...
Benchmarking ascii_from_string ...
Benchmarking ascii_to_string ...
Benchmarking erlang_unicode_from_gigantic_string ...
Benchmarking iso_from_gigantic_string ...
Benchmarking iso_from_string ...
Benchmarking iso_to_string ...
Calculating statistics...
Formatting results...

Name                                          ips        average  deviation         median         99th %
iso_to_string                            20804.63      0.00005 s    ±43.16%      0.00004 s      0.00012 s
ascii_to_string                          19755.02      0.00005 s    ±42.08%      0.00004 s      0.00012 s
iso_from_string                            472.55      0.00212 s    ±20.05%      0.00206 s      0.00329 s
ascii_from_string                          184.48      0.00542 s    ±16.68%      0.00530 s      0.00857 s
erlang_unicode_from_gigantic_string        0.0378        26.42 s     ±0.00%        26.42 s        26.42 s
ascii_from_gigantic_string                 0.0315        31.78 s     ±0.00%        31.78 s        31.78 s
iso_from_gigantic_string                   0.0285        35.14 s     ±0.00%        35.14 s        35.14 s

Comparison: 
iso_to_string                            20804.63
ascii_to_string                          19755.02 - 1.05x slower +0.00000 s
iso_from_string                            472.55 - 44.03x slower +0.00207 s
ascii_from_string                          184.48 - 112.78x slower +0.00537 s
erlang_unicode_from_gigantic_string        0.0378 - 549689.66x slower +26.42 s
ascii_from_gigantic_string                 0.0315 - 661141.81x slower +31.78 s
iso_from_gigantic_string                   0.0285 - 731031.74x slower +35.14 s

Memory usage statistics:

Name                                   Memory usage
iso_to_string                            0.00006 GB
ascii_to_string                          0.00006 GB - 1.00x memory usage +0 GB
iso_from_string                          0.00209 GB - 35.00x memory usage +0.00203 GB
ascii_from_string                        0.00834 GB - 140.00x memory usage +0.00829 GB
erlang_unicode_from_gigantic_string        14.94 GB - 250589.08x memory usage +14.94 GB
ascii_from_gigantic_string                 37.25 GB - 625006.00x memory usage +37.25 GB
iso_from_gigantic_string                   37.25 GB - 625002.50x memory usage +37.25 GB

**All measurements for memory usage were the same**
```
I suspect the [privat_appened](https://www.erlang.org/blog/more-optimizations/#appending-to-binaries-in-otp-26) optimization no longer applies.

Fritzelblitz commented:
I'm not quite sure if you can trust the RAM measurements, a few 100 kilobytes seems a bit low to me. But if you look at the memory consumption of the entire system, you can see a noticeable difference.
I suspect Benchee is not measuring binaries in the binary heap correctly.

hawkyre commented:
I've simplified the code as much as I could while keeping the benefits but I'm still getting the same benchmarks. Also, the main branch benchmarks for to_string were broken due to it returning a tuple, I've fixed those here as well. I've created a post on elixir forum asking for some help, so I hope we can get this sorted out soon. https://elixirforum.com/t/losing-binary-compilation-optimizations-when-moving-functions-to-dynamically-compiled-module/71565

hawkyre commented:
I've managed to fix it! Here are the benchmarks with the latest commit:

```
Name                                          ips        average  deviation         median         99th %
ascii_to_string                           2106.07      0.00047 s     ±8.93%      0.00047 s      0.00059 s
iso_to_string                             1697.06      0.00059 s     ±9.16%      0.00059 s      0.00075 s
iso_from_string                           1370.55      0.00073 s     ±6.42%      0.00073 s      0.00084 s
ascii_from_string                          310.03      0.00323 s    ±14.46%      0.00334 s      0.00396 s
ascii_from_gigantic_string                  0.133         7.51 s     ±0.00%         7.51 s         7.51 s
iso_from_gigantic_string                    0.108         9.28 s     ±0.00%         9.28 s         9.28 s
gigantic_iso_to_string                      0.104         9.65 s     ±0.00%         9.65 s         9.65 s
erlang_unicode_from_gigantic_string        0.0650        15.40 s     ±0.00%        15.40 s        15.40 s

Comparison: 
ascii_to_string                           2106.07
iso_to_string                             1697.06 - 1.24x slower +0.00011 s
iso_from_string                           1370.55 - 1.54x slower +0.00025 s
ascii_from_string                          310.03 - 6.79x slower +0.00275 s
ascii_from_gigantic_string                  0.133 - 15821.31x slower +7.51 s
iso_from_gigantic_string                    0.108 - 19543.09x slower +9.28 s
gigantic_iso_to_string                      0.104 - 20324.32x slower +9.65 s
erlang_unicode_from_gigantic_string        0.0650 - 32424.76x slower +15.40 s

Memory usage statistics:

Name                                   Memory usage
ascii_to_string                           171.88 KB
iso_to_string                             171.88 KB - 1.00x memory usage +0 KB
iso_from_string                           171.88 KB - 1.00x memory usage +0 KB
ascii_from_string                        6757.81 KB - 39.32x memory usage +6585.94 KB
ascii_from_gigantic_string                414.06 KB - 2.41x memory usage +242.19 KB
iso_from_gigantic_string                  171.88 KB - 1.00x memory usage +0 KB
gigantic_iso_to_string                    171.88 KB - 1.00x memory usage +0 KB
erlang_unicode_from_gigantic_string  15661817.87 KB - 91123.30x memory usage +15661645.99 KB

**All measurements for memory usage were the same**
```

Feel free to test it locally as well. 

I also added an extra test for gigantic to_string transformations.

I think it should be safe to commit now.

tallakt commented:
Thanks, it turned into 0.1.11
